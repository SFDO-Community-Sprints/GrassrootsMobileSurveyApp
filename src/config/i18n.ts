import * as Localization from 'expo-localization';
import i18n from 'i18n-js';

import ne from './locale/ne_IN.json';
import en from './locale/en_US.json';

import { DEFAULT_SF_LANGUAGE } from '../constants';
import { toShortLocale } from '../utility';

//expo-localization gives the result in a language-region format
const deviceLocale: string =
  Localization.locale.indexOf('-') !== 0 ? Localization.locale.split('-')[0] : Localization.locale;

i18n.defaultLocale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.locale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.fallbacks = true;
i18n.translations = { en, ne };

export default i18n;
