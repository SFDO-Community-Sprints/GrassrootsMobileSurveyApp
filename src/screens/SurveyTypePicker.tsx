import React, { useState, useEffect, useContext } from 'react';
import { View, FlatList, ImageBackground } from 'react-native';
import { Divider } from 'react-native-elements';
import { StackNavigationProp } from '@react-navigation/stack';

import { getAllRecordsWithCallback } from '../services/database/database';
import { ListItem } from '../components';

import LocalizationContext from '../context/localizationContext';
import {
  APP_THEME,
  BACKGROUND_IMAGE_SOURCE,
  BACKGROUND_STYLE,
  BACKGROUND_IMAGE_STYLE,
  L10N_PREFIX,
  DB_TABLE,
} from '../constants';
import { logger } from '../utility/logger';

import { StackParamList } from '../Router';

import { SQLiteRecordType } from '../types/sqlite';

type SurveyTypePickerNavigationProp = StackNavigationProp<StackParamList, 'SurveyTypePicker'>;

type Props = {
  navigation: SurveyTypePickerNavigationProp;
};

export default function SurveyTypePicker({ navigation }: Props) {
  const [recordTypes, setRecordTypes] = useState<Array<SQLiteRecordType>>([]);
  const { t } = useContext(LocalizationContext);

  useEffect(() => {
    const fetch = async () => {
      await getAllRecordsWithCallback(DB_TABLE.RECORD_TYPE, setRecordTypes);
    };
    fetch();
  }, []);

  return (
    <ImageBackground source={BACKGROUND_IMAGE_SOURCE} style={BACKGROUND_STYLE} imageStyle={BACKGROUND_IMAGE_STYLE}>
      <View>
        <FlatList
          keyExtractor={(item, index) => {
            return index.toString();
          }}
          data={recordTypes}
          renderItem={({ item }) => (
            <ListItem
              title={t(`${L10N_PREFIX.RecordType}${item.developerName}`)}
              onPress={() => {
                logger('DEBUG', 'SurveyTypePicker', item.recordTypeId);
                navigation.navigate('SurveyEditor', {
                  selectedLayoutId: item.layoutId,
                  selectedRecordTypeId: item.recordTypeId,
                });
              }}
            />
          )}
          ItemSeparatorComponent={() => <Divider style={{ backgroundColor: APP_THEME.APP_BORDER_COLOR }} />}
        />
      </View>
    </ImageBackground>
  );
}
