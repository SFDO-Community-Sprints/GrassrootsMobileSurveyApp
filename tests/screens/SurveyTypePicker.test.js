import React from 'react';
import { render } from '@testing-library/react-native';

import i18n from '../../src/config/i18n';
import LocalizationContext from '../../src/context/localizationContext';
import SurveyTypePicker from '../../src/screens/SurveyTypePicker';

const mockRecordTypes = [
  {
    recordTypeId: '0125h000000kNJMAA1',
    developerName: 'Test_1',
    label: 'Test 1',
    layoutId: '00h5h000000mTxOAAU',
    active: true,
    master: false,
  },
  {
    recordTypeId: '0125h000000kNJMAA2',
    developerName: 'Test_2',
    label: 'Test 2',
    layoutId: '00h5h000000mTxOAAU',
    active: true,
    master: false,
  },
];

jest.mock('../../src/services/database/metadata', () => ({
  getAllAvailableRecordTypes: jest.fn().mockImplementation(() => Promise.resolve(mockRecordTypes)),
}));

describe('<SurveyTypePicker />', () => {
  it('render', async () => {
    const locale = i18n.locale;
    const localizationContext = {
      t: jest.fn(),
      locale,
      setLocale: jest.fn(),
    };

    const screen = render(
      <LocalizationContext.Provider value={localizationContext}>
        <SurveyTypePicker navigation={global.mockedNavigation} />
      </LocalizationContext.Provider>
    );

    const menu = await screen.findByA11yRole('menubar');
    expect(menu.props.data).toHaveLength(2);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
