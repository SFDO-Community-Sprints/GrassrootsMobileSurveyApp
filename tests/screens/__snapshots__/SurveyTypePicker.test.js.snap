// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SurveyTypePicker /> render 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        Object {
          "resizeMode": "contain",
          "top": 857.7142857142858,
        },
      ]
    }
  />
  <View>
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      accessibilityRole="menubar"
      data={
        Array [
          Object {
            "active": true,
            "developerName": "Test_1",
            "label": "Test 1",
            "layoutId": "00h5h000000mTxOAAU",
            "master": false,
            "recordTypeId": "0125h000000kNJMAA1",
          },
          Object {
            "active": true,
            "developerName": "Test_2",
            "label": "Test 2",
            "layoutId": "00h5h000000mTxOAAU",
            "master": false,
            "recordTypeId": "0125h000000kNJMAA2",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "white",
                "minHeight": 44,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "space-around",
                  "marginLeft": 20,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2b2826",
                    "fontFamily": "SalesforceSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "paddingBottom": 5,
                    "paddingTop": 5,
                  }
                }
              />
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#dedcdb",
                "height": 0.5,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "white",
                "minHeight": 44,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "space-around",
                  "marginLeft": 20,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2b2826",
                    "fontFamily": "SalesforceSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "paddingBottom": 5,
                    "paddingTop": 5,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</View>
`;
